File: go/test/country_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/world-bank-data-sdk"
	"github.com/voxgig-sdk/world-bank-data-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestCountryEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Country(nil)
		if ent == nil {
			t.Fatal("expected non-nil CountryEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := countryBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "country." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set WORLDBANKDATA_TEST_COUNTRY_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		countryRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.country", setup.data)))
		var countryRef01Data map[string]any
		if len(countryRef01DataRaw) > 0 {
			countryRef01Data = core.ToMapAny(countryRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = countryRef01Data

		// LIST
		countryRef01Ent := client.Country(nil)
		countryRef01Match := map[string]any{}

		countryRef01ListResult, err := countryRef01Ent.List(countryRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, countryRef01ListOk := countryRef01ListResult.([]any)
		if !countryRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", countryRef01ListResult)
		}

		// LOAD
		countryRef01MatchDt0 := map[string]any{
			"id": countryRef01Data["id"],
		}
		countryRef01DataDt0Loaded, err := countryRef01Ent.Load(countryRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		countryRef01DataDt0LoadResult := core.ToMapAny(countryRef01DataDt0Loaded)
		if countryRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if countryRef01DataDt0LoadResult["id"] != countryRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func countryBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "country", "CountryTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read country test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse country test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"country01", "country02", "country03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("WORLDBANKDATA_TEST_COUNTRY_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"WORLDBANKDATA_TEST_COUNTRY_ENTID": idmap,
		"WORLDBANKDATA_TEST_LIVE":      "FALSE",
		"WORLDBANKDATA_TEST_EXPLAIN":   "FALSE",
		"WORLDBANKDATA_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["WORLDBANKDATA_TEST_COUNTRY_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["WORLDBANKDATA_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["WORLDBANKDATA_APIKEY"],
			},
			extra,
		})
		client = sdk.NewWorldBankDataSDK(core.ToMapAny(mergedOpts))
	}

	live := env["WORLDBANKDATA_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["WORLDBANKDATA_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/country_entity_test.rb
# Country entity test

require "minitest/autorun"
require "json"
require_relative "../WorldBankData_sdk"
require_relative "runner"

class CountryEntityTest < Minitest::Test
  def test_create_instance
    testsdk = WorldBankDataSDK.test(nil, nil)
    ent = testsdk.Country(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = country_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "country." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set WORLDBANKDATA_TEST_COUNTRY_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    country_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.country")))
    country_ref01_data = nil
    if country_ref01_data_raw.length > 0
      country_ref01_data = Helpers.to_map(country_ref01_data_raw[0][1])
    end

    # LIST
    country_ref01_ent = client.Country(nil)
    country_ref01_match = {}

    country_ref01_list_result, err = country_ref01_ent.list(country_ref01_match, nil)
    assert_nil err
    assert country_ref01_list_result.is_a?(Array)

    # LOAD
    country_ref01_match_dt0 = {
      "id" => country_ref01_data["id"],
    }
    country_ref01_data_dt0_loaded, err = country_ref01_ent.load(country_ref01_match_dt0, nil)
    assert_nil err
    country_ref01_data_dt0_load_result = Helpers.to_map(country_ref01_data_dt0_loaded)
    assert !country_ref01_data_dt0_load_result.nil?
    assert_equal country_ref01_data_dt0_load_result["id"], country_ref01_data["id"]

  end
end

def country_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "country", "CountryTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = WorldBankDataSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["country01", "country02", "country03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["WORLDBANKDATA_TEST_COUNTRY_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "WORLDBANKDATA_TEST_COUNTRY_ENTID" => idmap,
    "WORLDBANKDATA_TEST_LIVE" => "FALSE",
    "WORLDBANKDATA_TEST_EXPLAIN" => "FALSE",
    "WORLDBANKDATA_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["WORLDBANKDATA_TEST_COUNTRY_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["WORLDBANKDATA_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["WORLDBANKDATA_APIKEY"],
      },
      extra || {},
    ])
    client = WorldBankDataSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["WORLDBANKDATA_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["WORLDBANKDATA_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
